File: 구간합구하기4_11659.java
package Gold이상문제_정리;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class 구간합구하기4_11659 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N, M;

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        // N개의 수 , M개의 구간
        int[] nArr = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        int[] sArr = new int[N];

        // 구간합 구하기
        sArr[0] = nArr[0];
        for (int i = 1; i < N; i++) {
            sArr[i] = sArr[i - 1] + nArr[i];
        }

        StringBuilder str = new StringBuilder();
        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());
            int start = Integer.parseInt(st.nextToken()) - 1;
            int end = Integer.parseInt(st.nextToken()) - 1;

            int sum = 0;
            if (start == 0)
                sum = sArr[end];
            else
                sum = sArr[end] - sArr[start - 1];
            str.append(sum).append("\n");
        }
        System.out.println(str);
    }
}

File: 누적합/구간합구하기4_11659.kt
package 누적합

import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.StringTokenizer

/*
5 3
5 4 3 2 1
1 3
2 4
5 5
 */

private var N: Int = 0
private var M: Int = 0
private lateinit var A: Array<Int>
private lateinit var dp: Array<Int>
private val sb = StringBuilder()

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val (n, m) = br.readLine().split(" ").map { it.toInt() }
    N = n
    M = m
    A = Array(N + 1) { 0 }
    dp = Array(N + 1) { 0 }

    val st = StringTokenizer(br.readLine())
    repeat(N) { idx ->
        val i = idx + 1
        A[i] = st.nextToken().toInt()
        if (i == 0) dp[i] = A[i]
        else dp[i] = dp[i - 1] + A[i]
    }

    repeat(M) {
        val st = StringTokenizer(br.readLine())
        val s = st.nextToken().toInt()
        val e = st.nextToken().toInt()

        sb.append("${dp[e] - dp[s - 1]}\n")
    }

    print(sb)
}